src/test/java/com/google/sps/CommentRetrievalTest.java

```diff
@@ -0,0 +1,138 @@
+    // Return on the first call either 100 comments or the desired number.
+    // If there's no next page token, return nothing on the second call,
+    // Otherwise return the remainder of comments
+    CommentThreadListResponse firstResponse =
+        mockThreadListResponse(
+            Math.min(numExpectedComments, MAX_COMMENTS_PER_TOKEN),
+            (useNextPageToken) ? NEXT_PAGE_TOKEN : null,
+            FIRST_TOKEN_COMMENT);
+    if (useNextPageToken) {
+      CommentThreadListResponse secondResponse =
+          mockThreadListResponse(
+              Math.min(numExpectedComments - MAX_COMMENTS_PER_TOKEN, MAX_COMMENTS_PER_TOKEN),
+              NEXT_PAGE_TOKEN,
+              SECOND_TOKEN_COMMENT);
+      when(mockedCommentThreadList.execute()).thenReturn(firstResponse, secondResponse);
+    } else {
+      when(mockedCommentThreadList.execute()).thenReturn(firstResponse);
+    }
+
+    when(mockedYoutube.commentThreads().list(anyString())).thenReturn(mockedCommentThreadList);
+    commentRetriever = new YouTubeCommentRetriever(mockedYoutube);
+  }
+
+  // Creates list of comments of desired length.
+  // Page token is needed for iteration but its contents are irrelevant.
+  private CommentThreadListResponse mockThreadListResponse(
+      int desiredComments, String nextPageToken, String commentContent) {
+    CommentThread mockThread = dummyCommentThread(commentContent);
+    return new CommentThreadListResponse()
+        .setItems(Collections.nCopies(desiredComments, mockThread))
+        .setNextPageToken(nextPageToken);
+  }
+
+  private CommentThread dummyCommentThread(String comment) {
+    // This has to be done because for some reason it's impossible to directly change a comment's
+    // content
+    CommentThread mockedCommentThread = mock(CommentThread.class, RETURNS_DEEP_STUBS);
+    when(mockedCommentThread.getSnippet().getTopLevelComment().getSnippet().getTextDisplay())
+        .thenReturn(comment);
+    return mockedCommentThread;
+  }
+
+  // The simplest case: extract 100 comments from a video with more than 100 comments
+  @Test
+  public void testDefaultBehaviour() throws Exception {
+    setUpYouTubeMocks(HUNDRED);
+    List<CommentThread> comments = commentRetriever.retrieveComments(videoUrl, HUNDRED);
+    Assert.assertEquals(comments.size(), HUNDRED);
+  }
+
+  // Only load the comments that are in a video, without crashing
+  @Test
+  public void doesNotAttemptRetrieveExcess() throws Exception {
+    setUpYouTubeMocks(1);
+    List<CommentThread> comments = commentRetriever.retrieveComments(videoUrl, EXCESS_HUNDRED);
+    Assert.assertEquals(comments.size(), 1);
+  }
+
+  // Retrieve a specific amount of comments less than 100
+  @Test
+  public void retrievesSpecificNumComments() throws Exception {
+    setUpYouTubeMocks(12);
+    List<CommentThread> comments = commentRetriever.retrieveComments(videoUrl, 12);
+    Assert.assertEquals(comments.size(), 12);
+  }
```

Comment:
I would change the comment here, instead of saying 100, say max number of results in a request.

Reply:
Agreed